### ISSUE TITLE
Multilabel model requires a huge amount of RAM

### ISSUE THREAD

**Issue by piotrusieg** [catboost/catboost#2845]
**Catboost: 1.2.7**

Hi,
Lately I tried to switch my binary catboost model to multilabel model.
Training dataset which I use store around **14 GB.**
Model is defined as follows:

` 
 model = CatBoostClassifier(
        iterations=120,
        early_stopping_rounds=10,
        depth=4,
        loss_function='MultiLogloss',
        eval_metric='MultiLogloss',
        random_seed=12,
        cat_features=sparse_features,
        verbose=10,
        used_ram_limit='512gb'
    )`

Model has **134 features where 17 of them are sparse**, rest of them is dense. 
Model has **13** output classes.

However, training from that moment became impossible. Training requires so much RAM that it always exceed limit of 512gb. I tried to overcome it by adding limit of 512gb but it's still not working. I tried a lot of different parameters combinations but at the end none of them worked. 

The only thing that I noticed is seems that model is quite sensitive to the number of iterations defined during declaration even if I would never reach this iteration.

![Image](https://github.com/user-attachments/assets/4c09a992-f013-4f3e-b102-65effdd99ef4)

I repeat this experiment multiple times and conclusion was always the same: **the number of predefined iterations has a huge impact on RAM usage**. I know that multilabel model will definitely needs more RAM than binary model, however I am very surprised that when I defined number of iterations equal to 1 or 100 it works, but when it was 500, it exceed 512GB of RAM even before first iterations, as if only the fact of requesting 500 iterations crash the model. Model by default ended training around 120 iteration - I was able to test it when I decreased iterations to 200.
Binary model works perfectly every time - despite the fact of adding new features, adding more data etc. Binary model works in such manner that it assigns 1 when any of 14 classes occur.

I know that multilabel model can take more RAM but now it totally prevents me from training.
So I have 2 questions:

1. It is possible that in Catboost number of defined iterations has such impact on RAM even if it would never exceed first iteration?
2. What I can do to overcome issue with training multilabel model with catboost? I removed some of the least important features, limit number of iterations etc. etc. but I'm collecting more and more data and at some point I would need to overcome this. I don't know - should I maybe try to encode sparse features by myself, try different parameters or with different loss function? The Internet says that I should try with different parameters (I tried), limit number of output classes (I can't), remove some features (I did it), remove some sparse features(I can't). I would not ask if my dataset would be very huge - but how from 14 GB it needs to store more than 512 GB during training? I don't know.

Thank you!

**Comment by kraj2003** [comment]
1. Why does the number of iterations increase RAM even before the first iteration?
Yes, CatBoost does preallocate memory based on the number of iterations defined. It’s a known behavior, especially for multiclass problems, because:

Multiclass training stores a separate tree structure per class per iteration (when using One-vs-All or similar internal structures).

It tries to reserve memory buffers ahead of time for faster access during training.

Even if early stopping cuts training off early, CatBoost doesn’t reduce the memory it preallocates up front.

This is why even if it doesn't run 500 iterations, just defining it preloads massive RAM.

🔍 Memory complexity grows as: O(num_iterations × num_classes × depth × num_samples)

So with 13 classes, 500 iterations, and a depth of 4 — that’s huge already.

2. what to do?
A. Reduce iterations conservatively and rely more on early stopping
Define iterations=200 or 250 max. Let early stopping handle the rest. You can train a learning curve externally and see where overfitting starts.
B. Try using loss_function='MultiClass' instead of MultiLogloss
MultiClass is more memory-efficient and better optimized in some versions.

_loss_function='MultiClass'
eval_metric='MultiClass'_

C. Remove your custom label encoding (Let CatBoost handle categorical)
When you manually label encode, CatBoost loses the chance to optimize internally.

Try feeding the raw string/object columns directly as cat_features.

CatBoost is optimized for handling categorical features better than label encoding.
